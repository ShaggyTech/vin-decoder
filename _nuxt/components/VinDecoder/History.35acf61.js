(window.webpackJsonp=window.webpackJsonp||[]).push([[2],{330:function(e,t,n){},341:function(e,t,n){"use strict";n.r(t);var o=n(16),r=n(51),c=Object(o.defineComponent)({name:"VinDecoderHistory",setup:function(e,t){var n,c=t.root.$accessor;return{history:(n=c,{history:Object(o.computed)((function(){return Object(r.a)(n.history.history)}))}).history,clearHistory:function(e){return{addHistoryItem:function(t){e.history.addHistoryItem(t)},deleteHistoryItem:function(t){e.history.deleteHistoryItem(t)},clearHistory:function(){e.history.clearHistory()}}}(c).clearHistory}}}),l=n(49),h=n(52),d=n.n(h),v=n(156),f=n(319),y=n(317),x=(n(10),n(6),n(5),n(3),n(7),n(2)),O=n(114),j=n(95),m=n(1),P=n(9);function w(object,e){var t=Object.keys(object);if(Object.getOwnPropertySymbols){var n=Object.getOwnPropertySymbols(object);e&&(n=n.filter((function(e){return Object.getOwnPropertyDescriptor(object,e).enumerable}))),t.push.apply(t,n)}return t}var C=Object(P.a)(Object(O.a)("expansionPanels","v-expansion-panel","v-expansion-panels"),Object(j.b)("expansionPanel",!0)).extend({name:"v-expansion-panel",props:{disabled:Boolean,readonly:Boolean},data:function(){return{content:null,header:null,nextIsActive:!1}},computed:{classes:function(){return function(e){for(var i=1;i<arguments.length;i++){var source=null!=arguments[i]?arguments[i]:{};i%2?w(Object(source),!0).forEach((function(t){Object(x.a)(e,t,source[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(source)):w(Object(source)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(source,t))}))}return e}({"v-expansion-panel--active":this.isActive,"v-expansion-panel--next-active":this.nextIsActive,"v-expansion-panel--disabled":this.isDisabled},this.groupClasses)},isDisabled:function(){return this.expansionPanels.disabled||this.disabled},isReadonly:function(){return this.expansionPanels.readonly||this.readonly}},methods:{registerContent:function(e){this.content=e},unregisterContent:function(){this.content=null},registerHeader:function(e){this.header=e,e.$on("click",this.onClick)},unregisterHeader:function(){this.header=null},onClick:function(e){e.detail&&this.header.$el.blur(),this.$emit("click",e),this.isReadonly||this.isDisabled||this.toggle()},toggle:function(){var e=this;this.content&&(this.content.isBooted=!0),this.$nextTick((function(){return e.$emit("change")}))}},render:function(e){return e("div",{staticClass:"v-expansion-panel",class:this.classes,attrs:{"aria-expanded":String(this.isActive)}},Object(m.l)(this))}}),D=n(154),_=n(193),k=n(25),B=Object(P.a)(_.a,k.a,Object(j.a)("expansionPanel","v-expansion-panel-content","v-expansion-panel")).extend().extend({name:"v-expansion-panel-content",computed:{isActive:function(){return this.expansionPanel.isActive}},created:function(){this.expansionPanel.registerContent(this)},beforeDestroy:function(){this.expansionPanel.unregisterContent()},render:function(e){var t=this;return e(D.a,this.showLazyContent((function(){return[e("div",t.setBackgroundColor(t.color,{staticClass:"v-expansion-panel-content",directives:[{name:"show",value:t.isActive}]}),[e("div",{class:"v-expansion-panel-content__wrap"},Object(m.l)(t))])]})))}}),A=n(78),H=n(74);function $(object,e){var t=Object.keys(object);if(Object.getOwnPropertySymbols){var n=Object.getOwnPropertySymbols(object);e&&(n=n.filter((function(e){return Object.getOwnPropertyDescriptor(object,e).enumerable}))),t.push.apply(t,n)}return t}function I(e){for(var i=1;i<arguments.length;i++){var source=null!=arguments[i]?arguments[i]:{};i%2?$(Object(source),!0).forEach((function(t){Object(x.a)(e,t,source[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(source)):$(Object(source)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(source,t))}))}return e}var E=Object(P.a)(k.a,Object(j.a)("expansionPanel","v-expansion-panel-header","v-expansion-panel")).extend().extend({name:"v-expansion-panel-header",directives:{ripple:H.a},props:{disableIconRotate:Boolean,expandIcon:{type:String,default:"$expand"},hideActions:Boolean,ripple:{type:[Boolean,Object],default:!1}},data:function(){return{hasMousedown:!1}},computed:{classes:function(){return{"v-expansion-panel-header--active":this.isActive,"v-expansion-panel-header--mousedown":this.hasMousedown}},isActive:function(){return this.expansionPanel.isActive},isDisabled:function(){return this.expansionPanel.isDisabled},isReadonly:function(){return this.expansionPanel.isReadonly}},created:function(){this.expansionPanel.registerHeader(this)},beforeDestroy:function(){this.expansionPanel.unregisterHeader()},methods:{onClick:function(e){this.$emit("click",e)},genIcon:function(){var e=Object(m.l)(this,"actions")||[this.$createElement(A.a,this.expandIcon)];return this.$createElement(D.b,[this.$createElement("div",{staticClass:"v-expansion-panel-header__icon",class:{"v-expansion-panel-header__icon--disable-rotate":this.disableIconRotate},directives:[{name:"show",value:!this.isDisabled}]},e)])}},render:function(e){var t=this;return e("button",this.setBackgroundColor(this.color,{staticClass:"v-expansion-panel-header",class:this.classes,attrs:{tabindex:this.isDisabled?-1:null,type:"button"},directives:[{name:"ripple",value:this.ripple}],on:I(I({},this.$listeners),{},{click:this.onClick,mousedown:function(){return t.hasMousedown=!0},mouseup:function(){return t.hasMousedown=!1}})}),[Object(m.l)(this,"default",{open:this.isActive},!0),this.hideActions||this.genIcon()])}}),V=(n(330),n(120)),S=n(8);function R(object,e){var t=Object.keys(object);if(Object.getOwnPropertySymbols){var n=Object.getOwnPropertySymbols(object);e&&(n=n.filter((function(e){return Object.getOwnPropertyDescriptor(object,e).enumerable}))),t.push.apply(t,n)}return t}function M(e){for(var i=1;i<arguments.length;i++){var source=null!=arguments[i]?arguments[i]:{};i%2?R(Object(source),!0).forEach((function(t){Object(x.a)(e,t,source[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(source)):R(Object(source)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(source,t))}))}return e}var z=V.a.extend({name:"v-expansion-panels",provide:function(){return{expansionPanels:this}},props:{accordion:Boolean,disabled:Boolean,flat:Boolean,hover:Boolean,focusable:Boolean,inset:Boolean,popout:Boolean,readonly:Boolean,tile:Boolean},computed:{classes:function(){return M(M({},V.a.options.computed.classes.call(this)),{},{"v-expansion-panels":!0,"v-expansion-panels--accordion":this.accordion,"v-expansion-panels--flat":this.flat,"v-expansion-panels--hover":this.hover,"v-expansion-panels--focusable":this.focusable,"v-expansion-panels--inset":this.inset,"v-expansion-panels--popout":this.popout,"v-expansion-panels--tile":this.tile})}},created:function(){this.$attrs.hasOwnProperty("expand")&&Object(S.a)("expand","multiple",this),Array.isArray(this.value)&&this.value.length>0&&"boolean"==typeof this.value[0]&&Object(S.a)(':value="[true, false, true]"',':value="[0, 2]"',this)},methods:{updateItem:function(e,t){var n=this.getValue(e,t),o=this.getValue(e,t+1);e.isActive=this.toggleMethod(n),e.nextIsActive=this.toggleMethod(o)}}}),J=n(316),component=Object(l.a)(c,(function(){var e=this,t=e.$createElement,n=e._self._c||t;return e.history.length>0?n("v-card",{staticClass:"history-card mx-auto my-4",attrs:{"max-width":"600"}},[n("v-card-title",{staticClass:"history-card__title headline"},[e._v("\n    Search History\n    "),n("v-spacer"),e._v(" "),e.history.length>0?n("v-btn",{staticClass:"btn__clear-history mx-2 my-2",attrs:{"x-small":"",color:"deep-orange darken-4"},on:{click:function(t){return t.preventDefault(),e.clearHistory(t)}}},[e._v("\n      Clear History\n    ")]):e._e()],1),e._v(" "),n("v-expansion-panels",{attrs:{accordian:"",focusable:"",hover:"",flat:"",tile:""}},e._l(e.history,(function(t,o){return n("v-expansion-panel",{key:o},[n("v-expansion-panel-header",{staticClass:"title primary darken-4"},[e._v("\n        "+e._s(t.VIN)+"\n      ")]),e._v(" "),n("v-expansion-panel-content",[n("lazy-vin-decoder-results",{attrs:{"raw-results":t.results,color:"primary darken-4",tile:""},on:{"update:rawResults":function(n){return e.$set(t,"results",n)},"update:raw-results":function(n){return e.$set(t,"results",n)}}})],1)],1)})),1)],1):e._e()}),[],!1,null,null,null);t.default=component.exports;d()(component,{LazyVinDecoderResults:function(){return n.e(3).then(n.bind(null,342)).then((function(e){return e.default||e}))}}),d()(component,{VBtn:v.a,VCard:f.a,VCardTitle:y.d,VExpansionPanel:C,VExpansionPanelContent:B,VExpansionPanelHeader:E,VExpansionPanels:z,VSpacer:J.a})}}]);