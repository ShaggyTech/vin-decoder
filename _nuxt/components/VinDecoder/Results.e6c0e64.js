(window.webpackJsonp=window.webpackJsonp||[]).push([[3],{327:function(t,e,r){},328:function(t,e,r){},329:function(t,e,r){},338:function(t,e,r){"use strict";r.r(e);r(17);var n=r(16),o=r(316),l=Object(n.defineComponent)({name:"VinDecoderResults",inheritAttrs:!1,props:{rawResults:{required:!1,type:Object,default:null},loading:{required:!1,type:Boolean,default:!1},transition:{required:!1,type:String,default:"slide-y-transition"},tableHeight:{required:!1,type:[Number,String],default:300},maxWidth:{required:!1,type:[Number,String],default:600}},setup:function(t){var e=Object(n.ref)(null);return Object(n.onMounted)((function(){e.value=t.rawResults?Object(o.b)(t.rawResults):null})),Object(n.watch)((function(){return t.rawResults}),(function(t){e.value=t?Object(o.b)(t):null})),{results:e}}}),c=r(52),d=r(67),h=r.n(d),v=r(315),f=r(313),y=(r(10),r(6),r(5),r(3),r(7),r(2)),O=(r(327),r(27));function m(object,t){var e=Object.keys(object);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(object);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(object,t).enumerable}))),e.push.apply(e,r)}return e}function j(t){for(var i=1;i<arguments.length;i++){var source=null!=arguments[i]?arguments[i]:{};i%2?m(Object(source),!0).forEach((function(e){Object(y.a)(t,e,source[e])})):Object.getOwnPropertyDescriptors?Object.defineProperties(t,Object.getOwnPropertyDescriptors(source)):m(Object(source)).forEach((function(e){Object.defineProperty(t,e,Object.getOwnPropertyDescriptor(source,e))}))}return t}var x=O.a.extend({name:"v-divider",props:{inset:Boolean,vertical:Boolean},render:function(t){var e;return this.$attrs.role&&"separator"!==this.$attrs.role||(e=this.vertical?"vertical":"horizontal"),t("hr",{class:j({"v-divider":!0,"v-divider--inset":this.inset,"v-divider--vertical":this.vertical},this.themeClasses),attrs:j({role:"separator","aria-orientation":e},this.$attrs),on:this.$listeners})}}),_=(r(328),r(1)),w=r(9);function S(object,t){var e=Object.keys(object);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(object);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(object,t).enumerable}))),e.push.apply(e,r)}return e}var P=Object(w.a)(O.a).extend({name:"v-simple-table",props:{dense:Boolean,fixedHeader:Boolean,height:[Number,String]},computed:{classes:function(){return function(t){for(var i=1;i<arguments.length;i++){var source=null!=arguments[i]?arguments[i]:{};i%2?S(Object(source),!0).forEach((function(e){Object(y.a)(t,e,source[e])})):Object.getOwnPropertyDescriptors?Object.defineProperties(t,Object.getOwnPropertyDescriptors(source)):S(Object(source)).forEach((function(e){Object.defineProperty(t,e,Object.getOwnPropertyDescriptor(source,e))}))}return t}({"v-data-table--dense":this.dense,"v-data-table--fixed-height":!!this.height&&!this.fixedHeader,"v-data-table--fixed-header":this.fixedHeader},this.themeClasses)}},methods:{genWrapper:function(){return this.$slots.wrapper||this.$createElement("div",{staticClass:"v-data-table__wrapper",style:{height:Object(_.f)(this.height)}},[this.$createElement("table",this.$slots.default)])}},render:function(t){return t("div",{staticClass:"v-data-table",class:this.classes},[this.$slots.top,this.genWrapper(),this.$slots.bottom])}}),k=(r(46),r(32),r(35),r(42),r(19)),C=(r(329),r(186)),B=r(75);function $(object,t){var e=Object.keys(object);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(object);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(object,t).enumerable}))),e.push.apply(e,r)}return e}function E(t){for(var i=1;i<arguments.length;i++){var source=null!=arguments[i]?arguments[i]:{};i%2?$(Object(source),!0).forEach((function(e){Object(y.a)(t,e,source[e])})):Object.getOwnPropertyDescriptors?Object.defineProperties(t,Object.getOwnPropertyDescriptors(source)):$(Object(source)).forEach((function(e){Object.defineProperty(t,e,Object.getOwnPropertyDescriptor(source,e))}))}return t}var D=Object(w.a)(C.a,B.a,O.a).extend({name:"VSkeletonLoader",props:{boilerplate:Boolean,loading:Boolean,tile:Boolean,transition:String,type:String,types:{type:Object,default:function(){return{}}}},computed:{attrs:function(){return this.isLoading?this.boilerplate?{}:E({"aria-busy":!0,"aria-live":"polite",role:"alert"},this.$attrs):this.$attrs},classes:function(){return E(E({"v-skeleton-loader--boilerplate":this.boilerplate,"v-skeleton-loader--is-loading":this.isLoading,"v-skeleton-loader--tile":this.tile},this.themeClasses),this.elevationClasses)},isLoading:function(){return!("default"in this.$scopedSlots)||this.loading},rootTypes:function(){return E({actions:"button@2",article:"heading, paragraph",avatar:"avatar",button:"button",card:"image, card-heading","card-avatar":"image, list-item-avatar","card-heading":"heading",chip:"chip","date-picker":"list-item, card-heading, divider, date-picker-options, date-picker-days, actions","date-picker-options":"text, avatar@2","date-picker-days":"avatar@28",heading:"heading",image:"image","list-item":"text","list-item-avatar":"avatar, text","list-item-two-line":"sentences","list-item-avatar-two-line":"avatar, sentences","list-item-three-line":"paragraph","list-item-avatar-three-line":"avatar, paragraph",paragraph:"text@3",sentences:"text@2",table:"table-heading, table-thead, table-tbody, table-tfoot","table-heading":"heading, text","table-thead":"heading@6","table-tbody":"table-row-divider@6","table-row-divider":"table-row, divider","table-row":"table-cell@6","table-cell":"text","table-tfoot":"text@2, avatar@2",text:"text"},this.types)}},methods:{genBone:function(text,t){return this.$createElement("div",{staticClass:"v-skeleton-loader__".concat(text," v-skeleton-loader__bone")},t)},genBones:function(t){var e=this,r=t.split("@"),n=Object(k.a)(r,2),o=n[0],l=n[1];return Array.from({length:l}).map((function(){return e.genStructure(o)}))},genStructure:function(t){var e=[];t=t||this.type||"";var r=this.rootTypes[t]||"";if(t===r);else{if(t.indexOf(",")>-1)return this.mapBones(t);if(t.indexOf("@")>-1)return this.genBones(t);r.indexOf(",")>-1?e=this.mapBones(r):r.indexOf("@")>-1?e=this.genBones(r):r&&e.push(this.genStructure(r))}return[this.genBone(t,e)]},genSkeleton:function(){var t=[];return this.isLoading?t.push(this.genStructure()):t.push(Object(_.l)(this)),this.transition?this.$createElement("transition",{props:{name:this.transition},on:{afterEnter:this.resetStyles,beforeEnter:this.onBeforeEnter,beforeLeave:this.onBeforeLeave,leaveCancelled:this.resetStyles}},t):t},mapBones:function(t){return t.replace(/\s/g,"").split(",").map(this.genStructure)},onBeforeEnter:function(t){this.resetStyles(t),this.isLoading&&(t._initialStyle={display:t.style.display,transition:t.style.transition},t.style.setProperty("transition","none","important"))},onBeforeLeave:function(t){t.style.setProperty("display","none","important")},resetStyles:function(t){t._initialStyle&&(t.style.display=t._initialStyle.display||"",t.style.transition=t._initialStyle.transition,delete t._initialStyle)}},render:function(t){return t("div",{staticClass:"v-skeleton-loader",attrs:this.attrs,on:this.$listeners,class:this.classes,style:this.isLoading?this.measurableStyles:void 0},[this.genSkeleton()])}}),component=Object(c.a)(l,(function(){var t=this,e=t.$createElement,r=t._self._c||e;return r("None"!==t.transition?"v-"+t.transition:"div",{tag:"component",attrs:{"hide-on-leave":""}},[t.loading?r("v-skeleton-loader",{staticClass:"results-loader",attrs:{"max-width":t.maxWidth,type:"card-heading, text@7"}}):t.results&&t.rawResults?r("v-card",t._b({staticClass:"results-card",attrs:{"max-width":t.maxWidth,raised:""}},"v-card",t.$attrs,!1),[r("v-card-title",{staticClass:"results-card__title mx-auto px-2 title"},[r("v-divider"),t._v(" "),r("span",{staticClass:"mx-auto px-4"},[t._v("\n        "+t._s(t.results.ModelYear)+" "+t._s(t.results.Make)+" "+t._s(t.results.Model)+"\n        "+t._s(t.results.Series)+"\n      ")]),t._v(" "),r("v-divider")],1),t._v(" "),r("v-card-subtitle",{staticClass:"results-card__subtitle mx-auto px-2 title text-center"},[t._v("\n      "+t._s(t.results.VIN)+"\n      "),r("v-divider")],1),t._v(" "),r("v-card-text",[r("v-simple-table",{staticClass:"results-card__table",attrs:{"fixed-header":"",height:t.tableHeight,"max-width":t.maxWidth},scopedSlots:t._u([{key:"default",fn:function(){return[r("thead",[r("tr",[r("th",{staticClass:"text-left"},[t._v("Name")]),t._v(" "),r("th",{staticClass:"text-left"},[t._v("Value")])])]),t._v(" "),r("tbody",t._l(t.results,(function(e,n,o){return r("tr",{key:o},[r("td",[t._v(t._s(n))]),t._v(" "),r("td",[t._v(t._s(e))])])})),0)]},proxy:!0}])})],1)],1):t._e()],1)}),[],!1,null,null,null);e.default=component.exports;h()(component,{VCard:v.a,VCardSubtitle:f.b,VCardText:f.c,VCardTitle:f.d,VDivider:x,VSimpleTable:P,VSkeletonLoader:D})}}]);