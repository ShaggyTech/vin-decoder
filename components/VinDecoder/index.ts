/* Composition API */
import { ref } from '#imports';
/* Utilities */
import {
  fetchDecodeVinResults,
  getHistoryItemIndex,
  handleError,
} from '~/utils';
/* Types */
import {
  DecodeVinValuesExtendedResults,
  TypedVuexStore,
  Validator,
} from '~/types';

/* config for VIN input field's validation-provider */
export const VALIDATOR: Validator = {
  rules: { required: true, vin: true },
  immediate: false,
  vid: 'vin-input-validator',
  name: 'vin-input-validator',
  customMessages: {
    required: 'Please provide a VIN to decode.',
  },
};

export const setupRefs = () => ({
  alertMessage: ref<string | null>(null),
  loading: ref<boolean>(false),
  rawResults: ref<object | null>(null),
  validator: ref<Validator>(VALIDATOR),
  vin: ref<string | null>(null),
});

export const initializeComponent = (store: TypedVuexStore) => {
  /* All refs necessary for this component */
  const refs = { ...setupRefs() };

  /* Method called when a valid VIN is submitted */
  const getResults = async (
    vinValue: string | null
  ): Promise<DecodeVinValuesExtendedResults | void> => {
    /* Runtime type check */
    if (!vinValue || typeof vinValue !== 'string') {
      return;
    }

    /* Preparation */
    refs.loading.value = true;
    refs.alertMessage.value = null;

    /* Reduce API calls by returning previously decoded results from the history */
    const historyArray = store.history.history;
    const historyIndex = getHistoryItemIndex(vinValue, historyArray);
    if (historyIndex >= 0) {
      refs.loading.value = false;
      refs.rawResults.value = { ...historyArray[historyIndex].results };
      return {
        ...historyArray[historyIndex].results,
      };
    }

    /* Fetch the results using the nhtsa-api-wrapper */
    const results: DecodeVinValuesExtendedResults = await fetchDecodeVinResults(
      vinValue
    ).catch((err) => err);

    /* Handle any errors */
    if (results instanceof Error) {
      const errMsg =
        'Oops! It seems an error occurred when fetching data from the API';

      refs.rawResults.value = null;
      refs.alertMessage.value = errMsg;
      refs.loading.value = false;

      return handleError(results as Error, errMsg);
    }

    /* Extract and distribute the results */
    refs.rawResults.value = { ...results };
    store.history.addHistoryItem({
      VIN: results.VIN,
      results: { ...results },
    });

    /* Cleanup */
    refs.loading.value = false;
    return results;
  };

  return {
    ...refs,
    getResults,
  };
};
