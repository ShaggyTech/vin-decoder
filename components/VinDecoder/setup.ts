/* Composition API */
import { ref, Ref } from '@vue/composition-api';
/* Utilities */
import { fetchDecodeVinResults } from '@/utils/fetchDecodeVinResults';
import { handleError } from '@/utils/handleError';
/* Compositions */
import { getHistoryItemIndex } from '@/utils/getHistoryItemIndex';
/* Types */
import { DecodeVinValuesExtendedResults, Validator } from '@/types';
import { TypedVuexStore } from '@/store';

export type Refs = {
  alertMessage: Ref<string | null>;
  loading: Ref<boolean>;
  rawResults: Ref<object | null>;
  validator: Ref<Validator>;
  vin: Ref<string | null>;
};

/* config for VIN input field's validation-provider */
export const VALIDATOR: Validator = {
  rules: { required: true, vin: true },
  immediate: false,
  vid: 'vin-input-validator',
  name: 'vin-input-validator',
  customMessages: {
    required: 'Please provide a VIN to decode.'
  }
};

export const setupRefs = (): Refs => ({
  alertMessage: ref(null),
  loading: ref(false),
  rawResults: ref(null),
  validator: ref(VALIDATOR),
  vin: ref(null)
});

export const initializeComponent = (store: TypedVuexStore) => {
  /* All refs necessary for this component */
  const refs = { ...setupRefs() };

  /* Method called when a valid VIN is submitted */
  const getResults = async (
    vinValue: string | null
  ): Promise<DecodeVinValuesExtendedResults | void> => {
    /* Runtime type check */
    if (!vinValue || typeof vinValue !== 'string') {
      return;
    }

    /* Preparation */
    refs.loading.value = true;
    refs.alertMessage.value = null;

    /* Reduce API calls by returning previously decoded results from the history */
    const historyArray = store.history.history;
    const historyIndex = getHistoryItemIndex(vinValue, historyArray);
    if (historyIndex >= 0) {
      refs.loading.value = false;
      refs.rawResults.value = { ...historyArray[historyIndex].results };
      return {
        ...historyArray[historyIndex].results
      };
    }

    /* Fetch the results using the nhtsa-api-wrapper */
    const results: DecodeVinValuesExtendedResults = await fetchDecodeVinResults(
      vinValue
    ).catch(err => err);

    /* Handle any errors */
    if (results instanceof Error) {
      const errMsg =
        'Oops! It seems an error occurred when fetching data from the API';

      refs.rawResults.value = null;
      refs.alertMessage.value = errMsg;
      refs.loading.value = false;

      return handleError(results as Error, errMsg);
    }

    /* Extract and distribute the results */
    refs.rawResults.value = { ...results };
    store.history.addHistoryItem({
      VIN: results.VIN,
      results: { ...results }
    });

    /* Cleanup */
    refs.loading.value = false;
    return results;
  };

  return {
    ...refs,
    getResults
  };
};
