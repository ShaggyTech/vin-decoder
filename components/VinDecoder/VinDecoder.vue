<script lang="ts">
/* Composition API */
import { defineComponent } from '@vue/composition-api';
/* Components */
import { ValidationObserver } from 'vee-validate';
import BaseInputWithValidation from '@/components/base/BaseInputWithValidation.vue';
import VinResults from '@/components/VinDecoder/VinResults/VinResults.vue';

/* Component Setup */
import { initializeComponent } from '@/components/VinDecoder/setup';

export default defineComponent({
  name: 'VinDecoder',
  components: {
    BaseInputWithValidation,
    ValidationObserver,
    VinResults
  },

  setup(_, { root: { $accessor } }) {
    return {
      ...initializeComponent($accessor)
    };
  }
});
</script>

<template>
  <validation-observer v-slot="{ invalid }" slim>
    <v-card elevation="8" max-width="600" class="decoder-card mx-auto">
      <v-alert
        :value="alertMessage !== null"
        type="error"
        class="decoder-card__alert"
        dense
      >
        {{ alertMessage }}
      </v-alert>
      <v-card-title class="decoder-card__title headline">
        VIN Decoder
      </v-card-title>
      <v-card-text>
        <p>Enter your VIN to decode useful information about your vehicle.</p>
        <base-input-with-validation
          id="VinDecoderInput"
          v-model="vin"
          class="decoder-card__input"
          :validator="validator"
          counter="17"
          clearable
          outlined
          @click:clear="rawResults = null"
          @input="rawResults = null"
        />
      </v-card-text>
      <v-card-actions class="decoder-card__actions">
        <v-btn
          class="btn__get-results mx-auto my-4"
          color="success"
          :disabled="!vin || invalid"
          block
          @click.prevent="getResults(vin)"
        >
          Decode VIN
        </v-btn>
      </v-card-actions>
      <vin-results
        class="decoder-card__results"
        :raw-results.sync="rawResults"
        :loading.sync="loading"
        max-width="600"
        color="primary darken-4"
      />
    </v-card>
  </validation-observer>
</template>
