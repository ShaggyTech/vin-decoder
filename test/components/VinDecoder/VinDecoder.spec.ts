import { mount } from '@vue/test-utils';
import { VALIDATOR, setupRefs } from '@/components/VinDecoder/setup';
import VinDecoder from '@/components/VinDecoder/VinDecoder.vue';

const VALID_VIN = 'WVWDM7AJ1BW263846';

const refs = { ...setupRefs() };
const mockGetResults = jest.fn().mockResolvedValue(true);

const factory = (options: { [propName: string]: any } = {}) => {
  return mount(VinDecoder, {
    setup: () => ({
      ...refs,
      getResults: mockGetResults
    }),
    stubs: ['vin-results'],
    ...options
  });
};

describe('VinDecoder Component Tests', () => {
  test('base component state, matches snapshot', () => {
    const wrapper = factory();
    expect(wrapper.element).toMatchSnapshot();
  });

  test('base component state, submission button is disabled', () => {
    const wrapper = factory();
    expect(wrapper.find('.btn__get-results').attributes().disabled).toEqual(
      'disabled'
    );
  });

  test('alertMessage Ref exists and has default value of `null`', async () => {
    const wrapper = factory();
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$data.alertMessage).toBeDefined();
    expect(wrapper.vm.$data.alertMessage).toBe(null);
  });

  test('loading Ref exists and has default value of `false`', async () => {
    const wrapper = factory();
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$data.loading).toBeDefined();
    expect(wrapper.vm.$data.loading).toBe(false);
  });

  test('rawResults Ref exists and has default value of `null`', async () => {
    const wrapper = factory();
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$data.rawResults).toBeDefined();
    expect(wrapper.vm.$data.rawResults).toBe(null);
  });

  test('validator Ref exists and has default value equal to `VALIDATOR` constant', async () => {
    const wrapper = factory();
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$data.validator).toBeDefined();
    expect(wrapper.vm.$data.validator).toEqual(VALIDATOR);
  });

  test('vin Ref exists and has default value of `null`', async () => {
    const wrapper = factory();
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$data.vin).toBeDefined();
    expect(wrapper.vm.$data.vin).toBe(null);
  });

  test('vin Ref syncs with input value', async () => {
    const wrapper = factory();

    wrapper.find('#VinDecoderInput').setValue(VALID_VIN);

    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$data.vin).toBe(VALID_VIN);
  });

  test('submission button is clickable with valid VIN input', async () => {
    const wrapper = factory();

    wrapper.find('#VinDecoderInput').setValue(VALID_VIN);

    await wrapper.vm.$nextTick();
    wrapper.find('.btn__get-results').trigger('click');

    await wrapper.vm.$nextTick();
    expect(mockGetResults).toHaveBeenCalledWith(VALID_VIN);
  });

  test('an alert is shown if there is an error (while getting results)', async () => {
    const wrapper = factory();

    const alertElement = wrapper.find('.decoder-card__alert');
    const alertMessageText = 'Oops! Something went wrong.';

    expect(alertElement.isVisible()).toBe(false);

    wrapper.setData({ alertMessage: alertMessageText });

    await wrapper.vm.$nextTick();
    expect(alertElement.isVisible()).toBe(true);
    expect(alertElement.text()).toContain(alertMessageText);
  });
});
