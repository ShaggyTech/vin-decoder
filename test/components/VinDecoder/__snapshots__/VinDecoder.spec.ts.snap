// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VinDecoder Component Tests base component state, matches snapshot 1`] = `
<div
  class="decoder-card mx-auto v-card v-sheet theme--light elevation-8"
  style="max-width: 600px;"
>
  <div
    class="v-alert decoder-card__alert v-sheet theme--dark v-alert--dense error"
    role="alert"
    style="display: none;"
  >
    <div
      class="v-alert__wrapper"
    >
      <i
        aria-hidden="true"
        class="v-icon notranslate v-alert__icon material-icons theme--dark"
      >
        $error
      </i>
      <div
        class="v-alert__content"
      >
        
      
    
      </div>
    </div>
  </div>
   
  <div
    class="v-card__title decoder-card__title headline"
  >
    
      VIN Decoder
    
  </div>
   
  <div
    class="v-card__text"
  >
    <p>
      Enter your VIN to decode useful information about your vehicle.
    </p>
     
    <span
      class="decoder-card__input"
      id="validator"
    >
      <div
        class="v-input theme--light v-text-field v-text-field--enclosed v-text-field--outlined"
      >
        <div
          class="v-input__control"
        >
          <div
            class="v-input__slot"
          >
            <fieldset
              aria-hidden="true"
            >
              <legend
                style="width: 0px;"
              >
                <span>
                  â€‹
                </span>
              </legend>
            </fieldset>
            <div
              class="v-text-field__slot"
            >
              <input
                id="VinDecoderInput"
                type="text"
              />
            </div>
            <div
              class="v-input__append-inner"
            >
              <div
                class="v-input__icon v-input__icon--clear"
              >
                <button
                  aria-label="clear icon"
                  class="v-icon notranslate v-icon--disabled v-icon--link material-icons theme--light"
                  disabled="disabled"
                  type="button"
                >
                  $clear
                </button>
              </div>
            </div>
          </div>
          <div
            class="v-text-field__details"
          >
            <div
              class="v-messages theme--light"
            >
              <transition-group-stub
                class="v-messages__wrapper"
                name="message-transition"
                tag="div"
              />
            </div>
            <div
              class="v-counter theme--light"
            >
              0 / 17
            </div>
          </div>
        </div>
      </div>
    </span>
  </div>
   
  <div
    class="v-card__actions decoder-card__actions"
  >
    <button
      class="btn__get-results mx-auto my-4 v-btn v-btn--block v-btn--contained v-btn--disabled theme--light v-size--default"
      disabled="disabled"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        
        Decode VIN
      
      </span>
    </button>
  </div>
   
  <vin-results-stub
    class="decoder-card__results"
    color="primary darken-4"
    maxwidth="600"
    tableheight="300"
    transition="slide-y-transition"
  />
</div>
`;
